example/package-c/BUILD.bazel

```diff
@@ -0,0 +1,137 @@
+load("@rules_haskell//haskell:defs.bzl", "haskell_library", "haskell_test")
+
+# example for basic usage of the autodetect directive
+#
+# gazelle_haskell_modules:srcs: src/
+haskell_library(
+    name = "package-c",
+    compiler_flags = [
+        "-DVERSION_package_b=\"0.1.0.0\"",
+        "-XScopedTypeVariables",
+        "-lm",
+    ],
+    hidden_modules = ["PackageC.Other.B"],
+    version = "0.1.0.0",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//package-a",
+        "//package-a:package-a-v2",
+        "@stackage//:base",
+        "@stackage//:void",
+    ],
+)
+
+# example where we combine files in srcs with ones found by the directive
+#
+# gazelle_haskell_modules:srcs: more-src/
+haskell_library(
+    name = "package-c-explicit-srcs",
+    srcs = ["src/PackageC/Other/B.hs"],
+    compiler_flags = [
+        "-DVERSION_package_b=\"0.1.0.0\"",
+        "-XScopedTypeVariables",
+        "-lm",
+    ],
+    hidden_modules = [
+        "PackageC.Other.B",
+        "PackageC.Other.F",
+    ],
+    version = "0.1.0.0",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//package-a",
+        "//package-a:package-a-v2",
+        "@stackage//:base",
+        "@stackage//:void",
+    ],
+)
+
+# example where we have more than one folder to search
+#
+# gazelle_haskell_modules:srcs: src/ more-src/
+haskell_library(
+    name = "package-c-multiple-srcs",
+    compiler_flags = [
+        "-DVERSION_package_b=\"0.1.0.0\"",
+        "-XScopedTypeVariables",
+        "-lm",
+    ],
+    hidden_modules = [
+        "PackageC.Other.B",
+        "PackageC.Other.F",
+    ],
+    version = "0.1.0.0",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//package-a",
+        "//package-a:package-a-v2",
+        "@stackage//:base",
+        "@stackage//:void",
+    ],
+)
+
+# example where we have already autodetected some modules, but a new one has been added
+#
+# gazelle_haskell_modules:srcs: another-src/
+haskell_library(
+    name = "package-c-detect-new",
+    compiler_flags = [
+        "-DVERSION_package_b=\"0.1.0.0\"",
+        "-XScopedTypeVariables",
+        "-lm",
+    ],
+    hidden_modules = ["PackageC.Other.B"],
+    modules = [
+        ":package-c-detect-new.PackageC.Exposed.A",
+        ":package-c-detect-new.PackageC.Other.B",
```

Comment:
None of A or B uses modules that are autodiscovered. This means that the build won't fail if the modules aren't autodiscovered. Maybe change the module imports to make the build fail in this case?

Another approach would be to import the autodiscovered modules and the existing modules from test that depends on the library. This has the advantage of also ensuring that the old modules aren't removed by accident.

Other tests added in this PR also look affected by the same absence of checks.

Reply:
That is not to say that the pre-existing test wouldn't need a review in this respect as well :see_no_evil:

Reply:
Good point, I was focused more on giving examples, and not on enforcing that the build should only pass if autodiscovery works. I'll try to make them all pass iff the autodiscovery directive works correctly.

Reply:
I improved the not being tested situation, hopefully they all don't compile if the autodetect directive doesn't fire.